<img src=cenapad.jpg align=left><BR><BR><BR><BR><BR><BR>
<h1>Curso Introducción al Lenguaje Python<br>Programación Estructurada</h1>
<img src=python-logo.png align=left>

# 2.1 Primeros pasos hacia la programación¶

Por supuesto, podemos usar Python para tareas más complicadas que sumar dos y dos. Por ejemplo, podemos escribir una subsecuencia inicial de la serie de Fibonacci así:

In [ ]:
# Series de Fibonacci:
# la suma de dos elementos define el siguiente
a, b = 0, 1
while b < 10:
    print(b)
    a, b = b, a+b

In [ ]:
lista=[]
a, b = 0, 1

while b < 10:
    lista.append(b)
    a, b = b, a+b
    
print(lista)

Este ejemplo introduce varias características nuevas.
<ul>
<li>La primer línea contiene una asignación múltiple: las variables a y b toman en forma simultanea los nuevos valores 0 y 1. En la última linea esto se vuelve a usar, demostrando que las expresiones a la derecha son evaluadas antes de que suceda cualquier asignación. Las expresiones a la derecha son evaluadas de izquierda a derecha.</li><br>

<li>El bucle while se ejecuta mientras la condición (aquí: b < 10) sea verdadera. En Python, como en C, cualquier entero distinto de cero es verdadero; cero es falso. La condición también puede ser una cadena de texto o una lista, de hecho cualquier secuencia; cualquier cosa con longitud distinta de cero es verdadero, las secuencias vacías son falsas. La prueba usada en el ejemplo es una comparación simple.</li><br>

<li>El cuerpo del bucle tiene sangría: Python utiliza 4 espacios para diferencias que se encuentra en un bucle o función. El interpretador permite la utilziación de 4 espacios o tabulación (tab) pero no mezclar ambos tipos, pues puede generar errores en la interpretación.</li><br>

<li>La función print() escribe el valor de el o los argumentos que se le pasan. Difiere de simplemente escribir la expresión que se quiere mostrar en la forma en que maneja múltiples argumentos, cantidades en punto flotante, y cadenas. Las cadenas de texto son impresas sin comillas, y un espacio en blanco es insertado entre los elementos, así puedes formatear cosas de una forma agradable:</li>

</ul>

In [ ]:
i = 256 * 256
print('El valor de i es', i)

El parámetro nombrado end puede usarse para evitar el salto de linea al final de la salida, o terminar la salida con una cadena diferente:

In [ ]:
lista = []
a, b = 0, 1
while b < 1000:
    print(b, end=', ')
    lista.append(b)
    a, b = b, a+b

In [ ]:
for elemento in lista:
    print(elemento, end=", ")

Otros ejemplos de la utilización de <b>print</b>.

In [ ]:
print('Mercury', 'Venus', 'Earth', sep=', ')
print('Mars', 'Jupiter', 'Saturn', sep=', ')
print('Uranus', 'Neptune', 'Pluto', sep=', ')

In [ ]:
print('Mercury', 'Venus', 'Earth', sep=', ', end='\t\t')
print('Mars', 'Jupiter', 'Saturn', sep=', ', end='\t\t')
print('Uranus', 'Neptune', 'Pluto', sep=', ')

In [ ]:
print('Mercury', 'Venus', 'Earth', sep=', ')
print('Mars', 'Jupiter', 'Saturn', sep=', ')
print('Uranus', 'Neptune', 'Pluto', sep=', ')

In [ ]:
print('Printing in a Nutshell', end='\n * ')
print('Calling Print', end='\n * ')
print('Separating Multiple Arguments', end='\n * ')
print('Preventing Line Breaks')

Para más información sobre print visitar: https://realpython.com/python-print/

# 2.2. Condicionales: if, elif y else

La declaración if/elif/else funciona para controlar el fluxo de ejecución basandose en una o más condiciones. La sintaxis es la siguiente:

In [ ]:
#No ejecute está célula
if condicion:
    comando1
    comando2
    ...
    comandoN
elif condicion:
    comando1
    comando2
    ...
    comandoN
else:
    comando1
    comando2
    ...
    comandoN

Observa la <b>IDENTACIÓN</b> que define los bloques de código, a diferencia de otros lenguajes que usan delimitadores como llaves tales como "{" y "}" o bloques con "begin" y "end".<br>

Una buena práctica es usar <b>IDENTATION</b> con 4 espacios (primer nivel), 8 espacios (segundo nivel), etc.<br>

<b>NO SE DEBÉN USAR TABULACIONES</b> (debido a las convenciones de portabilidad y legibilidad), aunque están permitidas.<br>

<b>NUNCA se debe mezclar espacios con tabulaciones en el mismo programa.</b><br>

IDENTATION en Python es obligatorio y forma parte de la sintaxis. La ventaja de esta característica es la legibilidad del código. La mayoría de los editores de código o IDE de desarrollo facilitarán el trabajo de IDENTACIÓN colocando espacios automáticamente si se encuentra dentro del contexto de un bloque de código.

#### Ejemplo de uso del condicional if elif else:

In [ ]:
nombre = input("¿Cual es su nombre?:")
edad = int(input("¿Cual es su edad?:")) #Ingrese un número entero
grupo=""

if edad <= 12:
    grupo = "Infancia"
elif edad <= 18:
    grupo = "Adolescencia"
elif edad <= 26:
    grupo = "Juventud"
elif edad <= 65:
    grupo = "Adultez"
else:
    grupo = "Persona Mayor"

print(f"Nombre: {nombre} - Edad: {edad} - Pertenece al grupo {grupo}.")

La palabra reservada ‘elif‘ es una abreviación de ‘else if’, y es útil para evitar un sangrado excesivo. Una secuencia if ... elif ... elif ... sustituye las sentencias switch o case encontradas en otros lenguajes.

La declaración if/elif/else puede precisa tener  tener ninguún bloque elif, por ejemplo:

In [ ]:
edad=21
if edad<=17:
    print("Menor de edad")
else:
    print("Mayor de edad")

Es posible combinar condiones con otros operadores <b>and</b>, <b>not</b> y <b>or</b>.

In [ ]:
edad = 31
if edad <= 12:
    print("Infancia")
elif edad > 12 and edad < 21:
    print("Adolescente")
elif edad>=21:
    print("Adulto")

In [ ]:
x = int(input("Ingrese un número entero: "))
y = int(input("Ingrese otro número entero: "))
if x%2==1 and y%2==1:
    print("Ambos números son impares")
else:
    print("Por lo menos uno de los números es par")

In [ ]:
x = int(input("Ingrese un número entero: "))
y = int(input("Ingrese otro número entero: "))
if x%2==1 or y%2==1:
    print("Por lo menos uno delos números es impar")
else:
    print("Ambos números son pares")

In [ ]:
num = list(range(1,30))
for n in num:
    if n%3==0:
        print(f"{n} es divisible por 3")

Alternativamente, en lugar de imprimir en la pantalla, podemos crear una lista con los números divisibles por 3 y luego imprimirla

In [ ]:
num = list(range(1,30))
num_div3=[] #Iniciando una lista vacia (Esto es necesario para que podemos usar el método append)
for n in num:
    if n%3==0:
        num_div3.append(n)
print(num_div3)

### Ejercicio 1:

Dado tres puntos en el plano expresados como coordenadas (x, y) informar cuál es el que se encuentra más lejos del centro de coordenadas.

In [ ]:
import math
xs= []
ys = []

for i in range(0,3):
    xs.append(int(input(f"Ingrese pos x de coordenada {i+1}: ")))
    ys.append(int(input(f"Ingrese pos y de coordenada {i+1}: ")))

index = 0
dist = -1
for i in range(0,3):
    d=math.sqrt(xs[i]**2 + ys[i]**2)
    if d > dist:
        dist= d
        index = i
print(f"La coordenada más lejana es({xs[index]},{ys[index]})")


## Ejercicio 2:

Escribir un programa que almacene la cadena de caracteres contraseña en una variable, pregunte al usuario por la contraseña e imprima por pantalla si la contraseña introducida por el usuario coincide con la guardada en la variable sin tener en cuenta mayúsculas y minúsculas.

Sugerencia: utilizar la función lower.

<b>Salida deseada:</b><br>
La contaseña coincide<br>
<br>
La contraseña no coincide

In [ ]:
# Escriba el código aqui
guardar="contrasena"
contrasena=(input("Escriba la contrasena "))
if(contrasena.lower()=="contrasena"):
        print(" La contrasena coincide")
else:
        print(" la contrasena no coincide")
 

### Ejercicio 3:

Para tributar un determinado impuesto se debe ser mayor de 16 años y tener unos ingresos iguales o superiores a 2000 Bs mensuales. Escribir un programa que pregunte al usuario su edad y sus ingresos mensuales y muestre por pantalla si el usuario tiene que tributar o no.

<b>Salida deseada:</b><br>
Tienes que tributar<br>
<br>
No tienes que tributar

In [ ]:
edad=(int(input("Escriba su edad: ")))
ingresos=(int(input("Escriba sus ingresos: ")))
if(edad>=16 and ingresos>=2000):
    print("Tienes que tributar")
else:
    print("No tienes que tributar")

### Ejercicio 4:

Escriba un programa que pida dos números enteros y que calcule su división, escribiendo si la división es exacta o no. El programa debe tomar en cuenta que no se puede dividir por cero.

In [ ]:
num1=int(input("Escriba un numero"))
num2=int(input("Escriba otro numero"))

if num2 == 0:
    print("No se puede dividir entre 0")
elif num1 % num2 == 0:
    print("División exacta ", num1/num2)
elif num1 % num2 !=0:
    print("Division inexacta ", num1/num2)

### Ejercicio 5:

Escriba un programa que pida año, mes y día y retorne cuantos días han pasado o faltan a la fecha actual.

In [ ]:
from datetime import date

año= int(input("Ingrese año: "))
mes= int(input("Ingrese mes: "))
dia= int(input("Ingrese dia: "))
fecha = date(año,mes,dia)
ahora= date.today()
dif= abs(fecha-ahora).days

print(f"La diferencia de días es: {dif}")
    



### Ejercicio 6:

Escriba un programa que pida dos números enteros y que escriba si el mayor es múltiplo del menor. El programa debe avisar cuando se escriba valores negativos o nulos.

In [ ]:
numentero1=(int(input("Escriba un numero entero ")))
numentero2=(int(input("Escriba otro numero entero ")))
if numentero1 > numentero2:
    numentero1%numentero2 == 0
    print("El mayor es multiplo del menor")
elif numentero1 < numentero2:
    numentero1%numentero2 != 0
    print("No son multiplos")
else:
    print("Escribiste un valor negativo o nulo")


### Ejercicio 7:

Escriba un programa que pida los coeficientes de una ecuación de segundo grado (a x² + b x + c = 0) y escriba la solución.

Se recuerda que una ecuación de segundo grado puede no tener solución (solución con números complejos), tener una solución única, tener dos soluciones o que todos los números sean solución. Se recuerda que la fórmula de las soluciones cuando hay dos soluciones es x = (-b ± √(b2-4ac) ) / (2a)

En caso de no tener solución, hacer uso de la números complejos de Python.

Más información en:<br>
[1] https://relopezbriega.github.io/blog/2015/10/12/numeros-complejos-con-python/<br>
[2] https://pybonacci.org/2014/02/04/como-separar-parte-real-e-imaginaria-en-sympy/

<b>Salida deseada</b><br>
ECUACIÓN A X² + B X + C = 0<br>
Escriba el valor del coeficiente a: 2<br>
Escriba el valor del coeficiente b: -7<br>
Escriba el valor del coeficiente c: 3<br>
La ecuación tiene dos soluciones: 3.0 y 0.5

In [ ]:
import math
a = int(input("Coeficiente de A= "))
b = int(input("Coeficiente de B= "))
c = int(input("Coeficiente de C= "))
t=(b**2)-(4*a*c)
if(t>=0):
    sol = (-b + math.sqrt(t)) / (2*a)
    sol1 = (-b - math.sqrt(t)) / (2*a)
    if(sol==sol1):
        print("La ecuacion tiene una unica solucion",round(sol,2))
    else:
        print("La ecuacion tiene dos soluciones",round(sol,2)," y ",round(sol1,2))
        
else:
    sol = complex(round(-b/(2*a),2),math.sqrt(abs(t))/(2*a))
    sol1= complex(round(-b/(2*a),2),-math.sqrt(abs(t))/(2*a))
    print("La ecuacion tiene dos soluciones",sol," y ",sol1)

### Ejercicio 8:

Escriba un programa que pregunte primero si se quiere calcular el área de un triángulo o la de un círculo. Si se contesta que se quiere calcular el área de un triángulo (escribiendo T o t), el programa tiene que pedir entonces la base y la altura y escribir el área. Si se contesta que se quiere calcular el área de un círculo (escribiendo C o c), el programa tiene que pedir entonces el radio y escribir el área.

Se recuerda que el área de un triángulo es base por altura dividido por 2 y que el área de un círculo es Pi (aproximadamente 3,141592) por el radio al cuadrado. Utilizar 2 decimales para la salida.

Nota: Utilice la libreria math para obtener la función pi.

<b>Salida deseada:</b><br>
CÁLCULO DE ÁREAS<br>
Elija una figura geométrica:<br>
a) Triángulo<br>
b) Círculo<br>
¿Qué figura quiere calcular (Escriba T o C)? T<br>
Escriba la base: 3<br>
Escriba la altura: 5.5<br>
Un triángulo de base 3.0 y altura 5.0 tiene un área de 8.25<br>
<br>
CÁLCULO DE ÁREAS<br>
Elija una figura geométrica:<br>
a) Triángulo<br>
b) Círculo<br>
¿Qué figura quiere calcular (Escriba T o C)? C<br>
Escriba el radio: 2<br>
Un círculo de radio 2.0 tiene un área de 12.57<br>

In [ ]:
import math 
figura= input("¿Qué figura desea calcular? ")
if(figura=="T" or figura.lower()=="t"):  
    base= float(input("Escriba la base: "))
    altura= float(input("Escriba la altura: "))
    area= float((base*altura)/2)
    print(round(area,2))
elif(figura=="C" or figura.lower()=="c"):
    radio= float(input("Escriba un radio: "))
    area2=(math.pi*radio**2)
    print(round(area2,2))

# 2.3. La Sentencia for

La sentencia <b>for</b> en Python difiere un poco de lo que uno puede estar acostumbrado en lenguajes como C o Pascal. En lugar de siempre iterar sobre una progresión aritmética de números (como en Pascal) o darle al usuario la posibilidad de definir tanto el paso de la iteración como la condición de fin (como en C), la sentencia <b>for</b> de Python itera sobre los ítems de cualquier secuencia (una lista o una cadena de texto), en el orden que aparecen en la secuencia. Por ejemplo:

In [ ]:
# Midiendo cadenas de texto
palabras = ['gato', 'ventana', 'defenestrado']
for p in palabras:
    print(p, len(p)) 

Si necesitá modificar la secuencia sobre la que estás iterando mientras estás adentro del ciclo (por ejemplo para borrar algunos ítems), se recomienda que hagas primero una copia. Iterar sobre una secuencia no hace implícitamente una copia. La notación abajo es especialmente conveniente para esto:

In [ ]:
for p in palabras[:]:  # hace una copiade toda la lista
    if len(p) > 6:
        palabras.insert(0, p)

print(palabras)

Con for w in words:, el ejemplo intentaría crear una lista infinita, insertando defenestrado una y otra vez.<br>

Un bucle <b>for</b> funciona como un iterador. Un <b>for</b> se utiliza para realizar acciones sobre elementos de una secuencia o cualquier elemento iterable, como cadenas, listas y otros objetos iterables que veremos más adelante.

<b>Formato general para una declaración for</b>

for item in object:<br>
&nbsp;&nbsp;&nbsp;&nbsp;statements to do stuff

También es común usar un rango como iterable en un <b>for</b>

for i in range(0,10):<br>
&nbsp;&nbsp;&nbsp;&nbsp;hace alguna cosa con i

#### Ejemplo. El complemento de una secuencia de ADN se obtiene reemplazando todo As con Ts, Ts con As, Gs con Cs y Cs con Gs por Gs

In [ ]:
dna = "AGCAGGTGCGAGTAGCAGTAGCA"
dna

In [ ]:
#Transformar todas las A's en Ts, Ts en As, Gs en Cs y Cs en Gs
complemento=""  #Iniciando string vacia
for x in dna:  #A cada iteración en este bucle, x será una letra del dna
    if x=="G":  #Si x fuera G, concatena la string complemento con un C
        complemento = complemento + "C"
    elif x=="C": #Si x fuera C, concatena la string complemento con un T
        complemento = complemento + "G"
    elif x == "A":
        complemento = complemento + "T"
    elif x == "T":
        complemento = complemento + "A"
        
print(complemento)

<b>join</b> es un método de la clase string que concatena una cadena con las cadenas de una lista.<br><br>
<b>reversed</b> devuelve una lista o cadena invertida.<br><br>
Vamos a invertir una string y usar el método de join para formar una nueva cadena con el inverso complementario de la secuencia de DNA.

In [ ]:
reverso_complementar = ''.join(reversed(complemento)) #conatena la string vacia '' con el complemento invertido
print(f"DNA={dna}\nREVCOMP={reverso_complementar}")

In [ ]:
for i in range(len(dna)):
    print(f"Caracter de índice {i} en {dna[i]}")

In [ ]:
#El ciclo for anterior también se puede hacer usando la función enumerate
for i,x in enumerate(dna):
    print(f"Caracter de índice {i} é um {x}")

### Ejercicio 9:

Escriba un programa que pida un número de dados y muestre los valores de ese número de dados, al azar.

Nota. Utilizar la libreria random.
Más información en:
[1] https://stackoverflow.com/questions/3996904/generate-random-integers-between-0-and-9

<b>Salida deseada:</b><br><br>
TIRADA DE DADOS<br>
Número de dados: 0<br>
¡Imposible!<br>
<br>
TIRADA DE DADOS<br>
Número de dados:  5<br>
Dados: 3 5 1 1 3<br>

In [ ]:
import random 
var=int(input("Ingrese un número de dados "))
print ("Dados: ", end=" ")
for x in range(0,var): 
    print(random.randint(1,6),end=" ")

### Ejercicio 10:

Escriba un programa que pida un número de jugadores y tire un dado para cada jugador. El jugador que tire el mayor dado ganará el juego.

<b>Salida deseada:</b><br><br>
TIRADAS DE DADO<br>
Número de jugadores: 1<br>
¡Imposible!<br>
<br>
TIRADAS DE DADO<br>
Número de jugadores: 4<br>
Jugador 1: 3<br>
Jugador 2: 3<br>
Jugador 3: 5<br>
Jugador 4: 1<br>
El ganador es el Jugador 3.<br>
<br>
TIRADAS DE DADO<br>
Número de jugadores: 3<br>
Jugador 1: 4<br>
Jugador 2: 4<br>
Jugador 3: 1<br>
Los ganadores son el Jugador 1, Jugador 2.

In [ ]:
import random

i = int(input("Numero de Jugadores "))
if i <= 1:
    print("Imposible!")
jugadores= []

max = 0
for jugador in range (1,i):
    dado = random.randint(1,6)
    jugadores.append(dado)
    if dado >= max:
        max = dado
    print("Jugador", jugador, ": ", dado)

print("Los ganadores son: ")
for n in range(len(jugadores)):
    if jugadores[n] == max:
        print("Jugador", n+1)

### Ejercicio 11:

Escriba un programa que pida un número de dados y tire esa cantidad de dados. El primer jugador obtiene un punto por cada dado par. El segundo jugador obtiene un punto por cada dado impar. El jugador que saque más puntos, gana.

<b>Salida deseada:</b><br><br>
PARES Y NONES<br>
Número de dados: 0<br>
¡Imposible!<br>
<br>
PARES Y NONES<br>
Número de dados: 7<br>
Dados: 4 5 2 3 2 1 6<br>
Ha ganado el jugador de los pares.<br>
<br>
PARES Y NONES<br>
Número de dados: 8<br>
Dados: 1 4 1 3 5 6 6 2<br>
Han empatado.

In [1]:
import random
print("Tiradas de dado")
aux=int(input("Numero de dados: "))
guardar=[]
jugador1 = 0
jugador2 = 0
if (aux== 0):
    print("Imposible")
else: 
    print("Dados: ", end=" ")
    for x in range(0,aux):
        guardar.append(random.randint(1,6))
        print(guardar[x], end= " ")
    for n in range(0,aux):
        if(guardar[x]%2==0):
            jugador1=jugador1+1
        else:
            jugador2=jugador2+1
if (jugador1>jugador2):
    print("Ha ganado el jugador de los pares")
elif(jugador2>jugador1):
    print("")
    print("Ha ganado el jugador de los impares")
elif(jugador2 == jugador1):
    print("")
    print("Han empatado")

Tiradas de dado
Numero de dados: 9
Dados:  5 4 6 5 2 1 5 3 2 Ha ganado el jugador de los pares


### Ejercicio 12:

Escriba un programa que pida un número de dados y tire esa cantidad de dados. Si no salen dos dados iguales seguidos, el jugador gana. Si salen, pierde.

<b>Salida deseada:</b><br><br>
DADOS IGUALES<br>
Número de dados: 1<br>
¡Imposible!<br>
<br>
DADOS IGUALES<br>
Número de dados: 6<br>
Dados: 3 1 1 4 6 2<br>
El jugador ha perdido.<br>
<br>
DADOS IGUALES<br>
Número de dados: 4<br>
Dados: 4 2 5 3<br>
El jugador ha ganado.<br>

In [2]:
import random
print("TIRADAS DE DADO")
aux=int(input("Numero de dados: "))
valores=[1,2,3,4,5,6]
aux1=[]
r=0
repetidos=False
if(aux<=1):
    print("Imposible!")
else:
    print("Dados:",end=" ")
    for x in range(0,aux):
        aux1.append(random.randint(1, 6))
        print(aux1[x],end=" ") 
    for j in range(0,aux-1):
        r=aux1[j];
        if(r==(aux1[j+1])):
            repetidos=True
            break
        if(j==aux and aux1[j]==r):
            repetidos=True
if(repetidos):
    print("\nEl jugador ha perdido")
else:
    print("\nEl jugador ha ganado")

TIRADAS DE DADO
Numero de dados: 6
Dados: 1 5 5 6 1 6 
El jugador ha perdido


### Ejercicio 13:

Escribir un programa que pregunte al usuario una cantidad a invertir, el interés anual y el número de años, y muestre por pantalla el capital obtenido en la inversión cada año que dura la inversión.

<b>Salida deseada:</b><br><br>
¿Cantidad a invertir?  1000<br>
¿Interés porcentual anual?  10<br>
¿Años? 5<br>
Capital tras 1 años: 1100.00<br>
Capital tras 2 años: 1210.00<br>
Capital tras 3 años: 1331.00<br>
Capital tras 4 años: 1464.10<br>
Capital tras 5 años: 1610.51<br>

In [ ]:
inversion=int(input("Escriba cuando va a invertir: "))
interes=int(input("Cual será el interes porcentual anual?"))
anios= int(input("Años? "))
for n in range(0,anios):
    inversion=inversion+inversion*(interes/100)
    print("Capital tras", n+1, "anios: ", inversion)


# Ejercicio 14:

Escribir un programa que pida al usuario un número entero y muestre por pantalla un triángulo rectángulo ("* ") como el de más abajo, de altura el número introducido.

<b>Salida deseada:</b><br><br>
Introduce la altura del triángulo (entero positivo) 5<br>
*<br>
\* *<br>
\* * *<br>
\* * * *<br>
\* * * * * 

In [3]:
num1 = int(input("Introduce la altura del triangulo (entero positivo)"))
for i in range(1,num1+1):
    print("*"*i)

Introduce la altura del triangulo (entero positivo)5
*
**
***
****
*****


### Ejercicio 15:

Escribir un programa que pida al usuario un número entero y muestre por pantalla un triángulo rectángulo como el de más abajo.

<b>Salida deseada:</b><br><br>
Introduce la altura del triángulo (entero positivo):  9<br>

1 <br>
3 1 <br>
5 3 1 <br>
7 5 3 1 <br>
9 7 5 3 1

In [8]:
triangulo=int(input("Introduce un número impar (entero positivo) "))
lista=list(range(1,triangulo+1,2))
for i in range(0,len(lista)):
    print("")
    for l in range(i,-1,-1):
          print(lista[l], end= " ")

Introduce un número impar (entero positivo) 10

1 
3 1 
5 3 1 
7 5 3 1 
9 7 5 3 1 

### Ejercicio 16:

Escriba un programa que pida un número entero mayor que cero y que escriba sus divisores.

<b>Salida deseada:</b><br><br>
DIVISORES<br>
Escriba un número mayor que cero: -5<br>
¡Le he pedido un número entero mayor que cero!<br>
<br>
DIVISORES<br>
Escriba un número entero mayor que cero: 200<br>
Los divisores de 200 son 1 2 4 5 8 10 20 25 40 50 100 200

In [10]:
print("DIVISORES")
numero = int(input("Escriba un número entero mayor que cero: "))
if numero <= 0:
    print("¡Le he pedido un número entero mayor que cero!")
else:
    print(f"Los divisores de {numero} son ", end="")
    for i in range(1, numero // 2 + 1):
         if numero % i == 0:
            print(i, end=" ")
    print(f"{numero} ")

DIVISORES
Escriba un número entero mayor que cero: 521
Los divisores de 521 son 1 521 


### Ejercicio 17:

Escribir un programa que pida al usuario un número entero y muestre por pantalla un triángulo rectángulo como el de número de primos.

<b>Salida deseada:</b><br><br>
Introduce la altura del triángulo (entero positivo):  20<br>

1 <br>
3 1 <br>
5 3 1 <br>
7 5 3 1 <br>
11 7 5 3 1 <br>
13 11 7 5 3 1 <br>
17 13 11 7 5 3 1 <br>
19 17 13 11 7 5 3 1

In [11]:
primo=int(input("Introduce la altura del triángulo (entero positivo): "))
noprimos=list(range(2,primo+1))
primos=[2,3,5,7,11]
for x in range(11,len(noprimos)):
    if(not noprimos[x]%2==0 and not noprimos[x]%3==0 and not noprimos[x]%5==0 and not noprimos[x]%7==0 and not noprimos[x]%11==0):
        primos.append(noprimos[x])
for i in (range(0,len(primos))):
    print(" ")
    for l in range(i,-1,-1):
            print(primos[l],end=" ")

Introduce la altura del triángulo (entero positivo): 25
 
2  
3 2  
5 3 2  
7 5 3 2  
11 7 5 3 2  
13 11 7 5 3 2  
17 13 11 7 5 3 2  
19 17 13 11 7 5 3 2  
23 19 17 13 11 7 5 3 2 

### Ejercicio 18:

Escriba un programa que pida dos números enteros y escriba la suma de todos los enteros desde el primer número hasta el segundo.

<b>Salida deseada:</b><br><br>
SUMA ENTRE VALORES<br>
Escriba un número entero positivo: -5<br>
¡Le he pedido un número entero positivo!<br>
<br>
SUMA ENTRE VALORES<br>
Escriba un número entero positivo: 7<br>
Escriba un número entero mayor que 7: 7<br>
¡Le he pedido un número entero mayor que 7!<br>
<br>
SUMA ENTRE VALORES<br>
Escriba un número entero positivo: 30<br>
Escriba un número entero mayor que 30: 32<br>
30 + 31 + 32 = 93<br>

In [14]:
 print("SUMA ENTRE VALORES")
numero_1 = int(input("Escriba un número entero positivo: "))
if numero_1 <= 0:
    print("¡Le he pedido un número entero positivo!")
else:
    numero_2 = int(input(f"Escriba un número entero mayor que {numero_1}: "))
    if numero_2 <= numero_1:
         print(f"¡Le he pedido un número entero mayor que {numero_1}!")
    else:
        suma = 0
        for i in range(numero_1, numero_2 + 1):
             suma = suma + i
        print(f"La suma desde {numero_1} hasta {numero_2} es {suma}")
            

SUMA ENTRE VALORES
Escriba un número entero positivo: -1
¡Le he pedido un número entero positivo!


### Ejercicio 19:

Escriba un programa que pida un número entero mayor que cero y calcule su factorial.

<b>Salida deseada:</b><br><br>
FACTORIAL<br>
Escriba un número entero mayor que cero: -5<br>
¡Le he pedido un número entero mayor que cero!<br>
<br>
FACTORIAL<br>
Escriba un número entero mayor que cero: 5<br>
El factorial de 5 es 120<br>

In [16]:
print("FACTORIAL")
fact=(int(input("Escriba un número entero mayor que cero: ")))
acc=1
if(fact <=0):
    print("¡Le he pedido un número entero mayor que cero!")
else:
    for i in range(1,fact+1):
        acc=acc*i
    print(f"El factorial de {fact} es {acc}")

FACTORIAL
Escriba un número entero mayor que cero: 1
El factorial de 1 es 1


### Ejercicio 20:

Escribir un programa que pida al usuario un número entero (del 1 al 10) y genere la matriz caracol.

<b>Salida deseada:</b><br><br>

Introduce un número (entero positivo):  3<br>

1 2 3<br>
8 9 4<br>
7 6 5<br>

In [17]:
n = int(input("Introduce un número (entero positivo): "))
mat = []
for i in range(n):
    mat.append([])
    for j in range(n):
        mat[i].append(0)
    
fila = 0
col = 0
ruta = "der"

for i in range(1,(n*n)+1):
    mat[fila][col] = i
    
    if ruta == "der":
        if col+1 < n and mat[fila][col+1] == 0:
            col += 1
        else:
            fila += 1
            ruta = "abajo"
    elif ruta == "abajo":
        if fila+1 < n and mat[fila+1][col] == 0:
            fila += 1
        else:
            col -= 1
            ruta = "izq"
    elif ruta == "izq":
        if col-1 >= 0 and mat[fila][col-1] == 0:
            col -= 1
        else:
            fila -= 1
            ruta = "arriba"
    elif ruta == "arriba":
        if fila-1 >= 0 and mat[fila-1][col] == 0:
            fila -= 1
        else:
            col += 1
            ruta = "der"
            
for row in mat:
    for e in row:
        print(e,end=' ')
    print()

Introduce un número (entero positivo): 10
1 2 3 4 5 6 7 8 9 10 
36 37 38 39 40 41 42 43 44 11 
35 64 65 66 67 68 69 70 45 12 
34 63 84 85 86 87 88 71 46 13 
33 62 83 96 97 98 89 72 47 14 
32 61 82 95 100 99 90 73 48 15 
31 60 81 94 93 92 91 74 49 16 
30 59 80 79 78 77 76 75 50 17 
29 58 57 56 55 54 53 52 51 18 
28 27 26 25 24 23 22 21 20 19 


# 2.4. La Sentencia while

Una instrucción while prueba una condición inicial y comienza a ejecutar los comandos del bloque si la condición es verdadera. Mientras la condición permanezca verdadera, los comandos de bucle se ejecutan repetidamente.

Sintaxis:

mientras condición:

     comandos

#### Ejemplo de declaración while

In [1]:
i=0
while i < 10:
    i+=1
    print(i)

1
2
3
4
5
6
7
8
9
10


In [ ]:
# Iniciando con una lista vacia
nombres = []

# Defina una string llamada nuevo_nombre con cualquier valor diferente de "salir"
nuevo_nombre = ''

El ciclo while se ejecutará mientras el usuario ingrese nombres distintos a "salir", almacenando cada nombre ingresado en una lista.

In [ ]:
while nuevo_nombre != 'salir':
    # Pida que el usuario ingrese un nombre
    nuevo_nombre = input("Ingrese un nombre o la palabra salir: ")
    # Adiciona el nombre al final de la lista de nombres
    nombres.append(nuevo_nombre)

#Extraiga la palabra salir de la lista de names ("salir")
ultima = nombres.pop()
# Imprime la lista de nombres
print(nombres)

El control de bucles break, continue y pass permite:

In [ ]:
while True:
    line = input('Input > ')
    if line == 'done':
        break
    print(line)
print('Done!')

In [ ]:
while True:
    line = input('Input > ')
    if line[0] == '#':
        continue
    if line == 'done':
        break
    print(line)
print('Done!')

Más información:<br>
[1] https://www.pythonmania.net/es/2013/04/05/control-de-bucles-break-continue-y-pass/

### Ejercicio 21:

Escriba un programa que pida dos números enteros. El programa pedirá de nuevo el segundo número mientras no sea mayor que el primero. El programa terminará escribiendo los dos números.

<b>Salida deseada:</b><br><br>
NÚMERO MAYOR<br>
Escriba un número: 6<br>
Escriba un número mayor que 6: 6<br>
6 no es mayor que 6. Inténtelo de nuevo: 1<br>
1 no es mayor que 6. Inténtelo de nuevo: 8<br>
<br>
Los números que ha escrito son 6 y 8.

In [5]:
print("NÚMERO MAYOR")
primero = int(input("Escriba un número: "))
segundo = int(input(f"Escriba un número mayor que {primero}: "))

while segundo <= primero:
    segundo = int(input(f"{segundo} no es mayor que {primero}. Inténtelo de nuevo: "))
print()
print(f"Los números que ha escrito son {primero} y {segundo}.")

NÚMERO MAYOR
Escriba un número: 5
Escriba un número mayor que 5: 6

Los números que ha escrito son 5 y 6.


### Ejercicio 22:

Escriba un programa que pida números decimales mientras el usuario escriba número mayores que el anterior.


<b>Salida deseada:</b><br><br>
NÚMEROS MAYORES<br>
Escriba un número: 7<br>
Escriba un número mayor que 7.0: 1<br>
1.0 no es mayor que 7.0.<br>
<br>
NÚMEROS MAYORES<br>
Escriba un número: 9.3<br>
Escriba un número mayor que 9.3: 9.3<br>
9.3 no es mayor que 9.3.<br>
<br>
NÚMEROS MAYORES<br>
Escriba un número: 4.5<br>
Escriba un número mayor que 4.5: 5<br>
Escriba otro número mayor que 5.0: 7.3<br>
Escriba otro número mayor que 5.0: 2<br>
2.0 no es mayor que 4.5.

In [ ]:
#print("NÚMEROS MAYORES")
#primero = float(input("Escriba un número: "))
#segundo = float(input(f"Escriba un número mayor que {primero}: "))

#while segundo > primero:
#    segundo = float(input(f"Escriba un número mayor que {primero}: "))

#print()
#print(f"{segundo} no es mayor que {primero}.")


### Ejercicio 23:

Escriba un programa que pida un valor límite positivo y a continuación pida números hasta que la suma de los números introducidos supere el límite inicial.


<b>Salida deseada:</b><br><br>
HASTA EL LÍMITE<br>
Escriba el valor límite: -3<br>
El límite debe ser mayor que 0. Inténtelo de nuevo: 0<br>
El límite debe ser mayor que 0. Inténtelo de nuevo: 5.5<br>
Escriba un número: 7.2<br>
Ha superado el límite. La suma de los números introducidos es 7.2.<br>
<br>
HASTA EL LÍMITE<br>
Escriba el valor límite: 12<br>
Escriba un número: 5.2<br>
Escriba otro número: -1<br>
Escriba otro número: 9<br>
Ha superado el límite. La suma de los números introducidos es 13.2.

In [4]:
limite = float(input("Escriba el valor límite: "))
while limite <= 0:
     limite = float(input("El límite debe ser mayor que 0. Inténtelo de nuevo: "))

print()
numero = float(input("Escriba un número: "))
suma = numero

while suma < limite:
    numero = float(input("Escriba otro número: "))
    suma += numero
print()
print(f"Ha superado el límite. La suma de los números introducidos es {suma}.")
    


Escriba el valor límite: 3

Escriba un número: 2
Escriba otro número: 4

Ha superado el límite. La suma de los números introducidos es 6.0.


### Ejercicio 24:

Escriba un programa que calcule la descomposición en factores primos de un número.

<b>Salida deseada:</b><br><br>
DESCOMPOSICIÓN EN FACTORES PRIMOS<br>
Escriba un número entero mayor que 1: 500<br>
Descomposición en factores primos: 2 2 5 5 5<br>
<br>
DESCOMPOSICIÓN EN NÚMEROS PRIMOS<br>
Escriba un número entero mayor que 1: 521<br>
Descomposición en factores primos: 521<br>
<br>
DESCOMPOSICIÓN EN NÚMEROS PRIMOS<br>
Escriba un número entero mayor que 1: 1<br>
1 no es mayor que 1. Inténtelo de nuevo: 720<br>
Descomposición en factores primos: 2 2 2 2 3 3 5

In [5]:
print("DESCOMPOSICIÓN EN NÚMEROS PRIMOS")
numero = int(input("Escriba un número entero mayor que 1: "))
while numero <= 1:
    numero = int(input(f"{numero} no es mayor que 1. Inténtelo de nuevo: "))
    copia = numero
print("Descomposición en factores primos:", end="")
for i in range(2, numero + 1):
    while copia % i == 0:
        copia = copia // i
        print(f" {i}", end="")
print()

DESCOMPOSICIÓN EN FACTORES PRIMOS
Escriba un número entero mayor que 1: 5
Descomposición en factores primos: 5 

#### Ejercicio 25:

Escriba un programa que muestre números al azar del 1 al 6 mientras lo pida el usuario. El programa debe mostrar la suma de los números anteriores.

<b>Salida deseada:</b><br>

Juego del Azar<br>
Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar:<br>
Tirada: 2 - Suma: 2<br>
<br>
Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar:<br>
Tirada: 5 - Suma: 7<br>
<br>
Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar:<br>
Tirada: 4 - Suma: 11<br>
<br>
Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar: x<br>
Programa terminado.

In [ ]:
import random
print("Juego del Azar")
var=(input("Pulse Intro para lanzar el dado. Pulse otra tecla e Intro para terminar: "))
x = random.randint(1,6)
y=7

Juego del Azar


### Ejercicio 26:

Escriba un programa que pida un número del 1 al 6, un número de jugadores y tire un dado para cada jugador. El 1er jugador obtenga el mismo número será el ganador.

<b>Salida deseada:</b><br>

JUEGO DE LOS DADOS<br>
Número a acertar: 5<br>
Número de jugadores: 1<br>
¡Imposible!<br>
<br>
JUEGO DE LOS DADOS<br>
Número a acertar: 5<br>
Número de jugadores: 4<br>
Jugador 1: 3<br>
Jugador 2: 3<br>
Jugador 3: 5<br>
El ganador es el Jugador 3.<br>
<br>
JUEGO DE LOS DADOS<br>
Número a acertar: 5<br>
Número de jugadores: 3<br>
Jugador 1: 4<br>
Jugador 2: 4<br>
Jugador 3: 1<br>
No existe ganadores.

In [4]:
import random 
print("JUEGO DE LOS DADOS")
a=int(input("Número a acertar (1 al 6): "))
if(a>0 and a<=6):
    b=int(input("Número de jugadores"))
    ganador=False
    for x in range(1,b+1):
        jugador=random.randint(1,6)
        print(f"Jugador{x}: {jugador}")
        if (jugador == c):
            print(f"El ganador es el jugador {x}")
            ganador = True 
            break
        if(not ganador):
            print("No existen ganadores")
        else:
            print("No cumple con las condiciones para jugar")
              

JUEGO DE LOS DADOS
Número a acertar (1 al 6): 5
Número de jugadores3
Jugador1: 6


NameError: name 'c' is not defined

### Ejercicio 27:

Escribir un programa que genere <b>una cadena de 4 caracteres aleatorios</b> que usen minúsculas y números en una variable, pregunte al usuario si desea averiguar la cadena o utilizar fuerza bruta para conseguir la contraseña correcta. En caso de usar la opción de fuerza bruta, declarar la cantidad de iteraciones utilizadas para encontrar la contraseña por fuerza bruta.

<b>Salida deseada:</b><br>

QUEBRAR UNA CONTRASEÑA<br>

El programa generó una contraseña.<br>

¿Desea adivinar la contraseña (opción a) o desea utilizar fuerza bruta (opción fb)?: a<br>
Introduce la contraseña:  1234<br>
¡Incorrecto!

¿Desea adivinar la contraseña (opción a) o desea utilizar fuerza bruta (opción fb)?: a<br>
Introduce la contraseña:  pass<br>
¡Contraseña correcta!<br>
<br>
<br>
QUEBRAR UNA CONTRASEÑA<br>

El programa generó una contraseña.<br>

¿Desea adivinar la contraseña (opción a) o desea utilizar fuerza bruta (opción fb)?: a<br>
Introduce la contraseña:  cont<br>
¡Incorrecto!

¿Desea adivinar la contraseña (opción a) o desea utilizar fuerza bruta (opción fb)?: fb<br>
Se realizó 3465 iteraciones para encontrar la contraseña correcta: abc5<br>


In [ ]:
import random
print("QUEBRAR UNA CONTRASEÑA")
c = ''
for i in range(4):
    c = c + random.choice("abcdefghijklmnopqrstuvwxyz0123456789")
print("El programa generó una contraseña")
r = input("¿Desea adivinar la contraseña (opción a) o desea utilizar fuerza bruta (opción b)?: ")
if r == 'a':
    prueba = input("Introduce la contraseña: ")
    if prueba == c:
        print("¡Contraseña correcta!")
    else:
        print("¡Incorrecto!")
else:
    cont = 1
    posibilidades = ['a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z','0','1','2','3','4','5','6','7','8','9']
    ch1,ch2,ch3,ch4 = 0,0,0,0
    f = [posibilidades[ch1],posibilidades[ch2],posibilidades[ch3],posibilidades[ch4]]
    while ''.join(f) != c:
        cont += 1
        if ch4<len(posibilidades)-1:
            ch4 += 1
        elif ch3<len(posibilidades)-1:
            ch4 = 0
            ch3 += 1
        elif ch2<len(posibilidades)-1:
            ch4=0
            ch3=0
            ch2+=1
        elif ch1<len(posibilidades)-1:
            ch4=0
            ch3=0
            ch2=0
            ch1+=1      
        f = [posibilidades[ch1],posibilidades[ch2],posibilidades[ch3],posibilidades[ch4]]         
    print(f"Se realizó {cont} iteraciones para encontrar la contraseña correcta: {''.join(f)}")


### Ejercicio 28:

Escriba un programa para jugar a adivinar un número (el ordenador "piensa" el número y el usuario tiene que adivinarlo). El programa empieza pidiendo entre qué números está el número a adivinar, se "inventa" un número al azar y después el usuario va probando valores y el programa va diciendo si son demasiado grandes o pequeños.

<b>Salida deseada:</b><br>

Valor mínimo: 0<br>
Valor máximo: 100<br>
Intente adivinar un número entero entre 0 y 100.<br>
Escriba un número: 20<br>
¡Demasiado pequeño! Inténtelo de nuevo: 30<br>
¡Demasiado grande! Inténtelo de nuevo: 27<br>
¡Acertó! Le ha costado 3 intentos.

<b>Sugerencia:</b> puede perfeccionar este programa haciendo:

* que al principio el programa se asegure de que el valor máximo es superior al valor mínimo.
* que el programa detecte "trampas", por ejemplo, si al decir 25 le decimos "mayor" y al decir 26 le decimos "menor", el programa debe decirnos que estamos haciendo trampas y dejar de jugar.

In [ ]:
minimo=int(input("Valor mínimo: "))
maximo=int(input("Valor máximo: "))
if(minimo<maximo):
    adivi=random.randint(minimo, maximo)
    numero=int(input("Escriba un número: "))
    c=0
    while(not numero==adivi):
        if(numero<adivi):
            numero=int(input("¡Demasiado pequeño! Inténtelo de nuevo: "))
            c=c+1
        else:
            numero=int(input("¡Demasiado grande! Inténtelo de nuevo: "))
            c=c+1
    print(f"¡Acertó! Le ha costado {c} intentos.")
else:
    print("No cumple las condiciones para jugar")


### Ejercicio 29:

Escribir un programa que pida al usuario un número entero y muestre por pantalla un triángulo rectángulo de Fibonacci.

<b>Salida deseada:</b><br><br>
Introduce la altura del triángulo (entero positivo):  7<br>

1 <br>
1 1<br>
1 1 2<br>
1 1 2 3<br>
1 1 2 3 5<br>
1 1 2 3 5 8<br>
1 1 2 3 5 8 13<br>

In [2]:
n= int(input("Introduce la altura del triangulo (entero positivo): "))
lista=[]
n1,n2= 1,1
count = 0
while count < n:
    lista.append(n1)
    n=n1+n2
    n1=n2
    n2=n
    count+=1
    for n in range(0,len(lista)+1):
        for m in range(0,n,1):
            print(lista[m], end="")
        print("")


Introduce la altura del triangulo (entero positivo): 5

1


### Ejercicio 30:

Implemente un programa que sea capaz de calcular el resultado de aplicarla formula siguiente (n i)=n!/(i!\*(n-i)!). Para ello usted debe leer un numero por teclado que debe ser mayor e igual que cero teniendo en cuenta además que n tiene que ser necesariamente mayor que i y además calcular el factorial de un número.

In [3]:
n = int(input("Ingrese n: "))

def fact(x):
    r = 1
    for i in range(1,x+1):
        r *= i
    return r

if n < 0:
    print("n no puede ser menor a 0")
else:
    i = int(input("Ingrese i: "))
    if n <= i:
        print("n debe ser mayor a i")
    else:
        res = fact(n)/(fact(i)*fact(n-i))
        print(res)


Ingrese n: 5
Ingrese i: 5
n debe ser mayor a i
